libs/platform/ebpf_epoch.c

```diff
@@ -3,76 +3,62 @@
-// _ebpf_epoch_cpu_table[cpu_id].non_preemptible_epoch field.
-// 2) This field is read/written without explicit synchronization and can be old.
-// 3) _ebpf_epoch_update_cpu_entry periodically updates this on idle CPUs.
+// ebpf_epoch_exit:
+// First:
+// The current epoch is recorded in either a per thread or per CPU ebpf_epoch_state_t and it is marked as inactive.
+// For the per-thread case, the current CPUs per-thread table is first checked for an active epoch record. If not
+// found then each CPUs per-thread table is checked until the active record is found. This deals with the case where
+// a thread may switch CPUs between enter and exit.
 //
-// Preemptible invocations behavior:
-// 1) During ebpf_epoch_enter the value of _ebpf_current_epoch is captured and stored in the
-// _ebpf_epoch_cpu_table[cpu_id].thread_table[thread_id].entry_epoch field.
-// 2) During ebpf_epoch_exit the value of _ebpf_current_epoch is captured and stored in the
-// _ebpf_epoch_cpu_table[cpu_id].thread_table[thread_id].exit_epoch field.
-// 3) The _ebpf_epoch_cpu_table[cpu_id].thread_table is protected by _ebpf_epoch_cpu_table[cpu_id].thread_table_lock.
-// 4) If and only if entry_epoch > exit_epoch then the thread is actively executing between calls to ebpf_epoch_enter
-// and ebpf_epoch_exit.
+// Second:
+// Any entries in the per CPU free-list with epoch older than _ebpf_release_epoch are freed.
 //
-// Common behavior:
-// 1) Calls to ebpf_epoch_free mark the memory with _ebpf_current_epoch, atomically increment it and insert the memory
-// into the _ebpf_epoch_cpu_table[cpu_id].free_list while holding the _ebpf_epoch_cpu_table[cpu_id].free_list_lock.
-// 2) During ebpf_epoch_exit all entries in _ebpf_epoch_cpu_table[cpu_id].free_list where freed_epoch <
-// _ebpf_release_epoch are freed using ebpf_free.
-// 3) During ebpf_epoch_flush the minimum epoch is computed across the values in
-// _ebpf_epoch_cpu_table[*].non_preemptible_epoch and  _ebpf_epoch_cpu_table[*].thread_table[*].entry_epoch (for active
-// threads) and then written to _ebpf_release_epoch.
-// 4) ebpf_epoch_flush is called periodically by _ebpf_flush_timer.
-// 5) ebpf_epoch_flush also queues a non-preemptible work-item (_ebpf_epoch_update_cpu_entry) to all CPUs where
-// _ebpf_epoch_cpu_table[cpu_id].non_preemptible_epoch != _ebpf_current_epoch.
-// 6) _ebpf_epoch_update_cpu_entry calls ebpf_epoch_enter/ebpf_epoch_exit.
+// Third:
+// If the free-list still contains entries, the _ebpf_flush_timer is set (if not already set).
 //
-// Note:
-// CPU table entries aren't cleared on exit as we can't rely on memory ordering.
-// I.e., the thread doing the cleanup may have a stale view of the CPU table.
-// As long as the entries in the CPU table increase, this gives correct behavior.
+// ebpf_flush:
+// Compute the global lowest epoch across all active CPU and thread ebpf_epoch_state_t and set _ebpf_release_epoch.
```

Comment:
Since exit() only frees zombie objects on current cpu, and flush is not queueing DPCs and there is no updat_cpu_entry - wont this cause memory retention? When exit() finds there are pending zombies, it will re-arm the timer, timer updates release epoch - but then what?

Reply:
Consider the timer reaping free list when memory pressure builds up.

Reply:
Consider: adding a per cpu byte_allocated counter for diagnostics.

Reply:
#465 and #464